File: AC-tually Cold Enough/claude2.py
import sys

def main():
    data = sys.stdin.buffer.read().split()
    idx = 0
    N = int(data[idx]); D = int(data[idx+1]); K = int(data[idx+2])
    idx += 3

    H = [int(data[idx+i]) for i in range(N)]
    idx += N

    adj = [[] for _ in range(N)]
    for _ in range(N - 1):
        u = int(data[idx]); v = int(data[idx+1])
        idx += 2
        adj[u].append(v)
        adj[v].append(u)

    if K > N:
        print(-1)
        return

    parent = [-1] * N
    depth = [0] * N
    visited = [False] * N
    order = []
    stack = [0]
    visited[0] = True

    while stack:
        u = stack.pop()
        order.append(u)
        for w in adj[u]:
            if not visited[w]:
                visited[w] = True
                parent[w] = u
                depth[w] = depth[u] + 1
                stack.append(w)

    V = [0] * N
    V[0] = H[0]
    for u in order[1:]:
        val = H[u] + depth[u] * D
        V[u] = max(V[parent[u]], val)

    V.sort()
    print(V[K - 1])

main()

File: AC-tually Cold Enough/claude2.cpp
#include <bits/stdc++.h>
using namespace std;

int main(){
    int N; long long D; int K;
    scanf("%d %lld %d", &N, &D, &K);

    vector<long long> H(N);
    for(int i=0;i<N;i++) scanf("%lld", &H[i]);

    vector<vector<int>> adj(N);
    for(int i=0;i<N-1;i++){
        int u,v;
        scanf("%d %d", &u, &v);
        adj[u].push_back(v);
        adj[v].push_back(u);
    }

    if(K > N){
        printf("-1\n");
        return 0;
    }

    vector<long long> V(N, -1);
    vector<int> depth(N, 0);
    vector<int> parent(N, -1);

    vector<int> order;
    order.reserve(N);
    vector<int> stack_;
    stack_.push_back(0);
    vector<char> visited(N, 0);
    visited[0] = 1;

    while(!stack_.empty()){
        int u = stack_.back();
        stack_.pop_back();
        order.push_back(u);
        for(int w : adj[u]){
            if(!visited[w]){
                visited[w] = 1;
                parent[w] = u;
                depth[w] = depth[u] + 1;
                stack_.push_back(w);
            }
        }
    }

    V[0] = H[0];
    for(int idx = 1; idx < (int)order.size(); idx++){
        int u = order[idx];
        long long val = H[u] + (long long)depth[u] * D;
        V[u] = max(V[parent[u]], val);
    }

    sort(V.begin(), V.end());
    printf("%lld\n", V[K-1]);

    return 0;
}